Data/ByteString/Base64/Internal.hs

```diff
@@ -121,51 +121,46 @@ mkEncodeTable alphabet@(PS afp _ _) =
     table = B.pack $ concat $ [ [ix j, ix k] | j <- [0..63], k <- [0..63] ]
 
 -- | Decode a base64-encoded string.  This function strictly follows
--- the specification in
--- <http://tools.ietf.org/rfc/rfc4648 RFC 4648>.
+-- the specification in <http://tools.ietf.org/rfc/rfc4648 RFC 4648>.
+--
 -- This function takes the decoding table (for @base64@ or @base64url@) as
--- the first paramert.
+-- the first parameter.
+--
+-- For validation of padding properties, see note: $Validation
+--
 decodeWithTable :: Padding -> ForeignPtr Word8 -> ByteString -> Either String ByteString
 decodeWithTable _ _ (PS _ _ 0) = Right B.empty
-decodeWithTable padding decodeFP bs@(PS !fp !o !l) = unsafePerformIO $
+decodeWithTable padding decodeFP bs =
    case padding of
      Padded
-       | r == 1 -> err "Base64-encoded bytestring has invalid size"
-       | r /= 0 -> err "Base64-encoded bytestring required to be padded"
-       | otherwise -> go bs
+       | r == 0 -> unsafePerformIO $ go bs
+       | r == 1 -> Left "Base64-encoded bytestring has invalid size"
+       | otherwise -> Left "Base64-encoded bytestring is unpadded or has invalid padding"
      Don'tCare
-       | r == 0 -> go bs
-       | r == 2 -> go (B.append bs (B.replicate 2 0x3d))
-       | r == 3 -> go (B.append bs (B.replicate 1 0x3d))
-       | otherwise -> err "Base64-encoded bytestring has invalid size"
+       | r == 0 -> unsafePerformIO $ go bs
+       | r == 2 -> unsafePerformIO $ go (B.append bs (B.replicate 2 0x3d))
+       | r == 3 -> validateLastPad bs invalidPad $ go (B.append bs (B.replicate 1 0x3d))
+       | otherwise -> Left "Base64-encoded bytestring has invalid size"
      Unpadded
-       | r == 0 -> validateUnpadded (go bs)
-       | r == 2 -> validateUnpadded (go (B.append bs (B.replicate 2 0x3d)))
-       | r == 3 -> validateUnpadded (go (B.append bs (B.replicate 1 0x3d)))
-       | otherwise -> err "Base64-encoded bytestring has invalid size"
+       | r == 0 -> validateLastPad bs noPad $ go bs
+       | r == 2 -> validateLastPad bs noPad $ go (B.append bs (B.replicate 2 0x3d))
+       | r == 3 -> validateLastPad bs noPad $ go (B.append bs (B.replicate 1 0x3d))
```

Comment:
It's all written down in the `$Validation` note, but just to recap: 

Let `bs` be a bytestring of length `l`. Then the following properties hold: 

- `l == 0 mod 4`: Bytestring is padded and well-formed. This will always be the expected case for padded Base64 and Base64url values, or for unpadded Base64url values that happen to have a pre-encoded length multiple of 6. In any case, these will go through the standard decode routine, and any existing padding chars will be validated in the final quanta (see: `finalChunk`). 

- `l == 1 mod 4`: This is never a valid length for Base64 or Base64url-encoded values. The specification requires that unpadded length of the encoded string be either `l == 2 mod 4` or `l == 3 mod 4`. There will never be a valid unpadded input of length `l == 1 mod 4` as a result. 

- `l == 2 mod 4`: In this case, two padding chars must appear in the final quanta. If any additional padding chars exist in the string, then they will fail as final quanta, as we require the final four bytes (say, `(a b '=' '=')`) to have that `a /= '='` and `b /= '='`. Additionally pads will fail that clause of `finalCheck`. Thus, it's safe to add 2 padding chars to the end of a supposedly unpadded input, since the addition will never form a well-formed input if the unpadded string is already malformed. 

- `l == 3 mod 4`: This is the only tricky case. When inputs have this length, then we expect them to be of the form `(a b c '=')`. However, if the unpadded input has `'='` in the `c` position, it is possible that adding padding chars to the string "completes" the input in the sense that it forms a valid input where the unpadded fragment can be seen as a bytestring of length `l == 2 mod 4`. This could potentially be an attack vector, and constitutes a security risk. However, this is also easy to check, since, we only need to validate that the last char of an unpadded bytestring of length `l == 3 mod 4` is not `'='`. If any additional padding chars are present, then there is no risk that they will contribute to a well-formed input, since they will fail as final quanta in the `a` and `b` positions. So really, the requirement with padding bytestrings of length `l == 3 mod 4` is that they are of the form `(a b c '='), c /= '='` after padding.